File: convex/operations.ts
import { mutation, query } from "./_generated/server";
import { v } from "convex/values";

const makeId = () => crypto.randomUUID().replace(/-/g, "");

export const create = mutation({
  args: { clerk_user_id: v.string(), payload: v.any() },
  handler: async (ctx, args) => {
    const now = new Date().toISOString();
    const operation_id = typeof args.payload.operation_id === "string" ? args.payload.operation_id : makeId();
    const doc = {
      operation_id,
      clerk_user_id: args.clerk_user_id,
      feature: String(args.payload.feature ?? ""),
      status: String(args.payload.status ?? "queued"),
      queue_tier: String(args.payload.queue_tier ?? "free"),
      workload_pool: String(args.payload.workload_pool ?? "text_generation"),
      resource_scope: args.payload.resource_scope ?? {},
      idempotency_key: typeof args.payload.idempotency_key === "string" ? args.payload.idempotency_key : undefined,
      request_payload: args.payload.request_payload ?? {},
      response_payload: args.payload.response_payload ?? {},
      result_ref: args.payload.result_ref ?? undefined,
      sequence_no: typeof args.payload.sequence_no === "number" ? args.payload.sequence_no : undefined,
      request_id: typeof args.payload.request_id === "string" ? args.payload.request_id : undefined,
      error_code: typeof args.payload.error_code === "string" ? args.payload.error_code : undefined,
      error_message: typeof args.payload.error_message === "string" ? args.payload.error_message : undefined,
      enqueued_at: typeof args.payload.enqueued_at === "string" ? args.payload.enqueued_at : now,
      started_at: typeof args.payload.started_at === "string" ? args.payload.started_at : undefined,
      finished_at: typeof args.payload.finished_at === "string" ? args.payload.finished_at : undefined,
      created_at: typeof args.payload.created_at === "string" ? args.payload.created_at : now,
      updated_at: typeof args.payload.updated_at === "string" ? args.payload.updated_at : now,
    };
    await ctx.db.insert("operations", doc);
    return doc;
  },
});

export const get = query({
  args: { clerk_user_id: v.string(), operation_id: v.string() },
  handler: async (ctx, args) => {
    const row = await ctx.db
      .query("operations")
      .withIndex("by_user_operation", (q) =>
        q.eq("clerk_user_id", args.clerk_user_id).eq("operation_id", args.operation_id)
      )
      .first();
    if (!row) return null;
    const { _id, _creationTime, ...rest } = row;
    return rest;
  },
});

export const getByIdempotency = query({
  args: { clerk_user_id: v.string(), feature: v.string(), idempotency_key: v.string() },
  handler: async (ctx, args) => {
    const row = await ctx.db
      .query("operations")
      .withIndex("by_user_feature_idempotency", (q) =>
        q.eq("clerk_user_id", args.clerk_user_id).eq("feature", args.feature).eq("idempotency_key", args.idempotency_key)
      )
      .first();
    if (!row) return null;
    const { _id, _creationTime, ...rest } = row;
    return rest;
  },
});

export const update = mutation({
  args: { clerk_user_id: v.string(), operation_id: v.string(), payload: v.any() },
  handler: async (ctx, args) => {
    const row = await ctx.db
      .query("operations")
      .withIndex("by_user_operation", (q) =>
        q.eq("clerk_user_id", args.clerk_user_id).eq("operation_id", args.operation_id)
      )
      .first();
    if (!row) return null;
    const patch = {
      ...args.payload,
      updated_at: new Date().toISOString(),
    };
    await ctx.db.patch(row._id, patch);
    const { _id, _creationTime, ...rest } = { ...row, ...patch };
    return rest;
  },
});

File: backend/app/schemas/operations.py
"""Operation/job schemas used for async-capable workflows."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


OperationFeature = Literal[
    "chat_turn",
    "meal_plan_generate",
    "meal_plan_export_pdf",
    "recipe_generate",
    "recommendation_generate",
    "food_insight_analyze",
    "ingredient_check_analyze",
    "quiz_generate",
    "quiz_submit",
]

OperationStatus = Literal["queued", "running", "completed", "failed", "cancelled", "expired"]
OperationQueueTier = Literal["free", "plus", "pro"]


class OperationResultRef(BaseModel):
    resource_type: str
    resource_id: str
    session_id: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class OperationRecord(BaseModel):
    operation_id: str
    clerk_user_id: str
    feature: OperationFeature
    status: OperationStatus
    queue_tier: OperationQueueTier
    resource_scope: dict[str, str] = Field(default_factory=dict)
    workload_pool: str
    idempotency_key: str | None = None
    request_payload: dict[str, Any] = Field(default_factory=dict)
    response_payload: dict[str, Any] = Field(default_factory=dict)
    result_ref: OperationResultRef | None = None
    sequence_no: int | None = None
    request_id: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    enqueued_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class OperationSubmitRequest(BaseModel):
    feature: OperationFeature
    payload: dict[str, Any] = Field(default_factory=dict)
    idempotency_key: str | None = Field(default=None, min_length=8, max_length=200)
    resource_scope: dict[str, str] = Field(default_factory=dict)


class OperationResponse(BaseModel):
    operation: OperationRecord


class OperationStreamEvent(BaseModel):
    type: Literal["queued", "running", "completed", "failed", "heartbeat"]
    operation: OperationRecord
